esp/bindings/http/platform/httpbinding.cpp

```diff
@@ -685,7 +686,15 @@ bool EspHttpBinding::doAuth(IEspContext* ctx)
 {
     if(m_authtype.length() == 0 || stricmp(m_authtype.str(), "Basic") == 0)
     {
-        return basicAuth(ctx);
+        CumulativeTimer* timer = ctx->queryTraceSummaryCumulativeTimer("custom_fields.basicAuthTime");
+        timer->setLogLevel(LogNormal);
+        timer->setGroup(TXSUMMARY_GRP_ENTERPRISE);
```

Comment:
Should a timer previously configured for one group be assignable to another group? Consider a timer initially assigned to "core". I add another use of this timer but assign it to "enterprise". The timer will no longer be included in "core" output.

It seems like the request for a timer should include the intended group mask. If a timer exists with a matching mask, it is reused. If a timer exists with a different mask, we could either:

- Set the existing timer's mask to the union of the current and requested mask, and reuse the timer.
- Create a new timer using the requested mask and track time separately from the existing timer.
- Treat the mismatch as error to be fixed

The reuse of a timer should not blindly remove it from a previously assigned group.

A similar argument could also be made for log level, which is unchanged in this PR. Should timer reuse be able to exclude a previously recorded timed value by changing its log level after the fact? If yes, then should we be able to update every value this way?

Reply:
This feels analogous to the set() functions for the other types of entries- you can use them to change the logLevel, group or value of an existing entry. If you are explicitly calling the set() then you have the intention setting the field regardless of the previous value. Wasn't that the previous behavior of CumulativeTimers with respect to logLevel?

But of course that's just the convention and if you feel really strongly it should be different we can change it.

If we require logLevel and group in the request for a timer I think the only action we can take is your 3rd bullet above: treat mismatch as error. We can't add two entries with the same name, and I think that updating values as a side effect of a query call might be unexpected.

Reply:
They aren't exactly analogous. If I call set with an existing entry, the existing entry is destroyed. If I call setLogLevel or setGroup on a timer, the value is unchanged and only its visibility is impacted. This is why, aside from updating the header entries, there are very few (if any) occurrences of set.

Changing the visibility of an existing timer because a use of that timer wants non-default log level or group mask is also unexpected.

In most cases, this usage won't be problematic. My concern is that nothing can enforce a timer name won't be reused, and reused inappropriately. If we fail because of a mismatch it will be easier to find the problem.

Reply:
If anything I'm inclined to require group and logLevel in the `queryTimer()` call and they must match any existing entry. What if they don't match? Do I return null (in a new situation from before?) or throw a different exception? I think in this case that returning null would be preferred?

Reply:
The question is whether NULL was an expected outcome before. If yes, returning it here is fine. If no (except for the highly unlikely empty name scenario), I would think an exception is preferred to avoid dereferencing NULL.

Reply:
I feel like the scope of change for reworking the Timer interface is more than I can satisfactorily complete in the time I have for this PR. Although imperfect, I'll keep the `queryTimer()` and `updateTimer()` interfaces as they were prior to this PR. `queryTimer()` will throw exceptions for the new failure cases and I'll add the task of reviewing the interface to the next PR.

Reply:
Updated the timer interfaces to require level and group, and name matches found must also match on those to be returned. If name and group do not match an exception is thrown.